master/buildbot/secrets/providers/vault_hvac.py

```diff
@@ -0,0 +1,164 @@
+    """
+
+    def __init__(self, roleId, secretId):
+        self.roleId = roleId
+        self.secretId = secretId
+
+    def authenticate(self, client):
+        client.auth.approle.login(role_id=self.roleId, secret_id=self.secretId)
+
+
+class HvacKvSecretProvider(SecretProviderBase):
+    """
+    basic provider where each secret is stored in Vault KV secret engine
+    """
+
+    name = 'SecretInVaultKv'
+
+    def checkConfig(self, vaultServer=None, authenticator=None, secretsmount=None,
+                    apiVersion=2, pathDelimiter=':', pathEscape='\\'):
+        if not isinstance(vaultServer, str):
+            config.error("vaultServer must be a string while it is {}".format(type(vaultServer)))
+        if not isinstance(pathDelimiter, str) or len(pathDelimiter) > 1:
+            config.error("pathDelimiter must be a single character")
+        if not isinstance(pathEscape, str) or len(pathEscape) > 1:
+            config.error("pathEscape must be a single character")
+        if not isinstance(authenticator, HvacAuthenticator):
+            config.error("authenticator must be instance of HvacAuthenticator while it is {}"
+                         .format(type(authenticator)))
+        if apiVersion not in [1, 2]:
+            config.error("apiVersion {} is not supported".format(apiVersion))
+
+    def reconfigService(self, vaultServer=None, authenticator=None, secretsmount=None,
+                        apiVersion=2, pathDelimiter=':', pathEscape='\\'):
+        if secretsmount is None:
+            secretsmount = "secret"
+
+        self.secretsmount = secretsmount
+        self.pathDelimiter = pathDelimiter
+        self.pathEscape = pathEscape
+        self.authenticator = authenticator
+        self.apiVersion = apiVersion
+        if vaultServer.endswith('/'):
+            vaultServer = vaultServer[:-1]
+        self.client = hvac.Client(vaultServer)
+        self.client.secrets.kv.default_kv_version = apiVersion
+        return self
+
+    def escapedSplit(self, s):
+        """
+        parse and split string, respecting escape characters
+        """
+        ret = []
+        current = []
+        itr = iter(s)
+        for ch in itr:
+            if ch == self.pathEscape:
+                try:
+                    # skip the next character; it has been escaped and remove
+                    # escape character
+                    current.append(next(itr))
+                except StopIteration:
+                    pass
+            elif ch == self.pathDelimiter:
+                # split! (add current to the list and reset it)
+                ret.append(''.join(current))
+                current = []
+            else:
+                current.append(ch)
+        ret.append(''.join(current))
+        return ret
+
+    def hvacWrapRead(self, path):
+        if self.apiVersion == 1:
+            return self.client.secrets.kv.v1.read_secret(path=path, mount_point=self.secretsmount)
```

Comment:
This will involve network operations, won't it? We'll need to run this code from withing a thread pool thread so that whole Buildbot is not blocked while this function is running. See how `threads.deferToThread()` is used throughout the codebase. `def get()` will need to become `@defer.inlineCallbacks`, which is a sort of a Twisted's version of async/await.

Reply:
Yes, that's correct, this is network operation. I'll refactor it per your suggestion.

Reply:
you can see other code in buildbot, the code that is supposed to block shall be prefixed by thd_

So you can keep your code as is and make sure you use deferToThread and inlineCallbacks in the get() method.

Reply:
refactored in suggested way